optimization: add "lock transition" entity

eschava opened the pull request:
it's needed to replace plenty of edges between two nodes with one edge having multiple transition entities
it allows to dramatically speed up cycle detection step and improve memory consumed by found cycles
but on the other hand, some cycles potentially cannot be skipped on the "remove single-threaded locks", "remove gated cycles"

eschava committed:
optimization: add "lock transition" entity
it's needed to replace plenty of edges between two nodes with one edge having multiple transition entities
it allows to dramatically speed up cycle detection step and improve memory consumed by found cycles
but on the other hand, some cycles potentially cannot be skipped on the "remove single-threaded locks", "remove gated cycles"

toddlipcon commented:
Without studying the code quite a bit i'm not sure I understand this optimization.

Is this essentially "collapsing" multiple edges into a single edge in the graph that's used for analysis? eg if we have 'lockA' and 'lockB', and many different spots in the code where we have a lockA->lockB edge, we merge all those transitions into a single edge? Then because the graph has fewer edges (but same number of vertices) the cycle detection runs faster?

When you say "some cycles potentially cannot be skipped", do you mean that the end results are faster but less accurate? If so, maybe this should be an optional command line flag to run in a "fast mode" or something?

eschava commented:
You are absolutely right

This change could be improved to split cycle into several cycles while removing gated or single threaded cycles but it's hard to predict if it will be still useful after such change

About optional mode. Looks like it's the best option. I'll try to analyze how to change code to have both options in single codebase

eschava commented:
I think the best way to aggregate both options is having transitions as the collection inside edge but in "full" mode that collection will always be "Singleton list"
Agree?

toddlipcon commented:
That seems reasonable, so long as the extra memory/perf overhead isn't too expensive.

eschava committed:
optimization: add "lock transition" entity
it's needed to replace plenty of edges between two nodes with one edge having multiple transition entities
it allows to dramatically speed up cycle detection step and improve memory consumed by found cycles
but on the other hand, some cycles potentially cannot be skipped on the "remove single-threaded locks", "remove gated cycles"

eschava commented:
Done
